airflow/migrations/versions/7b2661a43ba3_taskinstance_keyed_to_dagrun.py

```diff
@@ -337,7 +337,7 @@ def downgrade():
         batch_op.drop_constraint('task_instance_pkey', type_='primary')
         batch_op.alter_column('execution_date', existing_type=dt_type, existing_nullable=True, nullable=False)
         batch_op.alter_column(
-            'dag_id', existing_type=string_id_col_type, existing_nullable=True, nullable=True
+            'dag_id', existing_type=string_id_col_type, existing_nullable=True, nullable=False
```

Comment:
```suggestion
            'dag_id', existing_type=string_id_col_type, existing_nullable=False, nullable=True
```

Shouldn't it be like this? Upgrade went from nullable -> nonnullable, we switch it back in downgrade?

Reply:
MSSQL doesn't like it:
```
sqlalchemy.exc.ProgrammingError: (pyodbc.ProgrammingError) ('42000', "[42000] [Microsoft][ODBC Driver 17 for SQL Server][SQL Server]Cannot define PRIMARY KEY constraint on nullable column in table 'task_instance'. (8111) (SQLExecDirectW)")
[SQL: ALTER TABLE task_instance ADD CONSTRAINT task_instance_pkey PRIMARY KEY (dag_id, task_id, execution_date)]
```

Reply:
If you leave this as not-nullable you haven't correctly reversed the migration.

You'll need to drop the indexes/constraints and re-create them for MSSQL to make this a full downgrade.

Reply:
MSSQL refused to drop the index `task_instance_pkey` it created on the table. I have tried different things but none is working. 
One thing that revealed what's happening to me was this `op.execute('alter table task_instance DROP INDEX task_instance_pkey')` :
```python
    batch_op.drop_constraint('task_instance_dag_run_fkey', type_='foreignkey')
    batch_op.drop_constraint('task_instance_pkey', type_='primary')
    batch_op.drop_index('ti_dag_run')
    batch_op.drop_index('ti_state_lkp')
    batch_op.drop_column('run_id')
    batch_op.alter_column('execution_date', existing_type=dt_type, existing_nullable=True, nullable=False)
    batch_op.alter_column(
        'dag_id', existing_type=string_id_col_type, existing_nullable=False, nullable=True
    )
    batch_op.drop_constraint('task_instance_pkey', type_='primary')
    if dialect_name == 'mssql':
        op.execute('alter table task_instance DROP INDEX task_instance_pkey')
        op.execute(
            """
            CREATE UNIQUE CLUSTERED INDEX task_instance_pkey
            ON task_instance (dag_id, task_id, execution_date) 

           """
        )
    else:
        batch_op.create_primary_key('task_instance_pkey', ['dag_id', 'task_id', 'execution_date'])
```
This gives the error that ```The operation 'ALTER TABLE DROP INDEX' is supported only with memory optimized tables. (10785) (SQLExecDirectW)")
[SQL: alter table task_instance DROP INDEX task_instance_pkey]```

Without the  `op.execute('alter table task_instance DROP INDEX task_instance_pkey')` it gives this error:

```Cannot create more than one clustered index on table 'task_instance'. Drop the existing clustered index 'task_instance_pkey' before creating another```

I don't know why the constraint `task_instance_pkey` is not dropping despite that I called `batch_op.drop_constraint('task_instance_pkey', type_='primary')` twice.

I will appreciate any helps cc: @jedcunningham @ashb

Reply:
What do you get with `op.execute('DROP INDEX task_instance_pkey ON task_instance')` instead? (the alter table version is only for memory optimized tables, so that error makes sense)

Looking at a couple other migrations, try dropping the index first too? Something like:
```
batch_op.drop_index('idx_task_instance')
batch_op.drop_constraint('task_instance_pkey', type_='primary')
```

Reply:
Tried all of the above but it didn't work. Tried a couple of other things but none worked. At last what I found out is, for `MSSQL` the dag_id was not nullable before the upgrade that made `dag_id` not nullable in other DBs. It created a clustered, unique key and made sure those columns were not nullable.

I don't think it's possible to make the `dag_id` nullable for mssql and still create a primary key out of it.

Reply:
> What do you get with `op.execute('DROP INDEX task_instance_pkey ON task_instance')` instead?

It gave this error:
```
An explicit DROP INDEX is not allowed on index 'task_instance.task_instance_pkey'. It is being used for PRIMARY KEY constraint enforcement
```

Reply:
Maybe the primary key is not named `task_instance_pkey` so the previous `drop_constraint` did not work as intended? I seem to recall we needed to do introspection on MSSQL in some migrations to get the correct primary key constraint name.

Reply:
Yes. The previous primary key was not named task_instance_pkey, it was named `PK__tas.....` can't remember the full name but what I found out is, the columns(dag_id, task_id, and execution_date ) were all non-nullable in mssql before the upgrade that created `task_instance_pkey`

Reply:
So if dag_id wasn't nullable before, we don't need this line then?

Reply:
If we do need to do it we probably have to have something like this https://stackoverflow.com/a/13949856/439189 (via `op.execute(sql)`

Reply:
```
irflow> select * from sysobjects where xtype = 'pk'; 
ime: 0.455s
--------------------------------+------------+-------
 name                           | id         | xtype 
--------------------------------+------------+-------
 PK__ab_permi__3213E83FCD8A3437 | 18099105   | PK    
 PK__ab_user__3213E83F5089F06C  | 98099390   | PK    
 PK__ab_user___3213E83FCC95BF08 | 194099732  | PK    
 PK__ab_regis__3213E83FF6637652 | 274100017  | PK    
 PK__sensor_i__3213E83F6142D05D | 338100245  | PK    
 PK__rendered__84D206CDB99874E2 | 370100359  | PK    
 PK__trigger__3213E83F7975A370  | 434100587  | PK    
 PK__session__3213E83FBB563C67  | 594101157  | PK    
 PK__log_temp__3213E83F2527E89D | 642101328  | PK    
 task_instance_pkey             | 690101499  | PK    
 PK__task_map__D8B3F9C159134C40 | 754101727  | PK    
 xcom_pkey                      | 818101955  | PK    
 alembic_version_pkc            | 901578250  | PK    
 PK__connecti__3213E83FE3E5EA5F | 933578364  | PK    
 PK__dag__0B56E00C25C44109      | 965578478  | PK    
 PK__dag_pick__3213E83F30AB8F06 | 997578592  | PK    
 PK__import_e__3213E83FA421F6BB | 1029578706 | PK    
 PK__job__3213E83F78CBFDF3      | 1061578820 | PK    
 PK__log__3213E83F04614680      | 1093578934 | PK    
 PK__slot_poo__3213E83F18DD869F | 1157579162 | PK    
 PK__variable__3213E83F885FB0EB | 1285579618 | PK    
 PK__dag_run__3213E83FB0BBA8B8  | 1413580074 | PK    
 PK__task_fai__3213E83F8E51376D | 1477580302 | PK    
 PK__task_res__3213E83F483158EF | 1653580929 | PK    
 PK__serializ__0B56E00C5F11BDE8 | 1717581157 | PK    
 PK__sla_miss__9BEABD047C43D94E | 1781581385 | PK    
 PK__dag_tag__A254411AF6010029  | 1813581499 | PK    
 PK__dag_code__786E9E46FCFC28E0 | 1893581784 | PK    
 PK__ab_permi__3213E83F68146FAF | 1941581955 | PK    
 PK__ab_view___3213E83FEEF1828D | 1989582126 | PK    
 PK__ab_role__3213E83FEBED9A0F  | 2037582297 | PK    
 PK__ab_permi__3213E83FCBF84AD3 | 2085582468 | PK    
--------------------------------+------------+-------
```

Almost nothing has a name PKey :(

Reply:
> So if dag_id wasn't nullable before, we don't need this line then?

It was nullable in Postgres, MySQL and SQLite but not in mssql

Reply:
Here's the table before the upgrade that made `dag_id` non-nullable:

```
airflow> sp_help task_instance                                                  
Time: 1.870s (a second)
+---------------+---------+------------+-------------------------+
| Name          | Owner   | Type       | Created_datetime        |
|---------------+---------+------------+-------------------------|
| task_instance | dbo     | user table | 2022-02-24 08:08:16.490 |
+---------------+---------+------------+-------------------------+
 
+----------------------+-----------+------------+----------+--------+---------+-
| Column_name          | Type      | Computed   | Length   | Prec   | Scale   | 
|----------------------+-----------+------------+----------+--------+---------+-
| task_id              | varchar   | no         | 250      |        |         | 
| dag_id               | varchar   | no         | 250      |        |         | 
| execution_date       | datetime2 | no         | 8        | 26     | 6       | 
| start_date           | datetime2 | no         | 8        | 26     | 6       | 
| end_date             | datetime2 | no         | 8        | 26     | 6       | 
| duration             | float     | no         | 8        | 53     | NULL    | 
| state                | varchar   | no         | 20       |        |         | 
| try_number           | int       | no         | 4        | 10     | 0       | 
| hostname             | varchar   | no         | 1000     |        |         | 
| unixname             | varchar   | no         | 1000     |        |         | 
| job_id               | int       | no         | 4        | 10     | 0       | 
| pool                 | varchar   | no         | 256      |        |         | 
| queue                | varchar   | no         | 256      |        |         | 
| priority_weight      | int       | no         | 4        | 10     | 0       | 
| operator             | varchar   | no         | 1000     |        |         | 
:
| queued_dttm          | datetime2 | no         | 8        | 26     | 6       | 
:
| pid                  | int       | no         | 4        | 10     | 0       | 
:
| max_tries            | int       | no         | 4        | 10     | 0       | 
:
| executor_config      | varbinary | no         | -1       |        |         | 
:
| pool_slots           | int       | no         | 4        | 10     | 0       | 
:
| queued_by_job_id     | int       | no         | 4        | 10     | 0       | 
:
| external_executor_id | varchar   | no         | 250      |        |         | 
| trigger_id           | int       | no         | 4        | 10     | 0       | 
:
| trigger_timeout      | datetime  | no         | 8        |        |         | 
:
| next_method          | varchar   | no         | 1000     |        |         | 
:
| next_kwargs          | nvarchar  | no         | -1       |        |         | 
+----------------------+-----------+------------+----------+--------+---------+-
:
 
:
+-----------------------------+--------+-------------+-----------------------+
:
| Identity                    | Seed   | Increment   | Not For Replication   |
:
|-----------------------------+--------+-------------+-----------------------|
:
| No identity column defined. | NULL   | NULL        | NULL                  |
:
+-----------------------------+--------+-------------+-----------------------+
 
:
+-------------------------------+
:
| RowGuidCol                    |
:
|-------------------------------|
| No rowguidcol column defined. |
:
+-------------------------------+
 
:
+-----------------------------+
:
| Data_located_on_filegroup   |
:
|-----------------------------|
:
| PRIMARY                     |
:
+-----------------------------+
 
:
+--------------------------------+----------------------------------------------
:
| index_name                     | index_description                            
:
|--------------------------------+----------------------------------------------
:
| PK__task_ins__9BEABD042B51623C | clustered, unique, primary key located on PRI
| ti_dag_date                    | nonclustered located on PRIMARY              
:
| ti_dag_state                   | nonclustered located on PRIMARY              
| ti_job_id                      | nonclustered located on PRIMARY              
:
| ti_pool                        | nonclustered located on PRIMARY              
| ti_state                       | nonclustered located on PRIMARY              
:
| ti_state_lkp                   | nonclustered located on PRIMARY              
:
| ti_trigger_id                  | nonclustered located on PRIMARY              
+--------------------------------+----------------------------------------------
:
 
:
+------------------------------+--------------------------------+---------------
| constraint_type              | constraint_name                | delete_action 
:
|------------------------------+--------------------------------+---------------
:
| DEFAULT on column max_tries  | DF__task_inst__max_t__5AEE82B9 | (n/a)         
:
| DEFAULT on column pool_slots | DF__task_inst__pool___17036CC0 | (n/a)         
:
| PRIMARY KEY (clustered)      | PK__task_ins__9BEABD042B51623C | (n/a)         
:
| FOREIGN KEY                  | task_instance_trigger_id_fkey  | Cascade       
:
|                              |                                |               
+------------------------------+--------------------------------+---------------
:
 
:
+-----------------------------------------------------------------+
:
| Table is referenced by foreign key                              |
:
|-----------------------------------------------------------------|
:
| airflow.dbo.task_reschedule: task_reschedule_dag_task_date_fkey |
:
+-----------------------------------------------------------------+
```
So `dag_id` wasn't nullable in MSSQL before the upgrade